Python3 l10n

darkdreamingdan opened the pull request:
This PR refactors the Python l10n scripts.  I've generally modernized them to be Python3 ready and written unit tests around them.  Opening this for feedback.

darkdreamingdan committed:
Port MTA scripts to python 3 and add python dependencies

darkdreamingdan committed:
upgraded NSIS script to python3 with tests

darkdreamingdan committed:
Remove old code

darkdreamingdan committed:
Add more tests.  Delete legacy files

darkdreamingdan committed:
Merge branch 'master' into python3_l10n

patrikjuvonen commented on utils/localization/README.md:
```diff
@@ -1 +1,36 @@
+These scripts will run in a native Python 3 environment.  You can install from the website or appstore as appropriate.
+
+# Setting up development environment
+For testing, a Conda environment is provided.  This is the reccomended distribution for installing Python on Windows.
```
Nit: Typo fix.
```suggestion
For testing, a Conda environment is provided.  This is the recommended distribution for installing Python on Windows.
```

patrikjuvonen commented on utils/localization/environment.yml:
```diff
@@ -0,0 +1,7 @@
+  - python=3.9
+  - pip
+  - pip:
+    - pytest
```
Nit: Fix missing newline at the end of the file.
```suggestion
    - pytest

```

patrikjuvonen commented on utils/localization/tests/test_build_gettext_catalog_nsi.py:
```diff
@@ -0,0 +1,69 @@
+    positions = nsi.get_newline_positions(sample_text)
+    line_no = nsi.get_line_number_from_position(positions, input_pos)
+    # Then
+    assert line_no == expected_line
```
Nit: Fix missing newline at the end of the file.
```suggestion
    assert line_no == expected_line

```

patrikjuvonen commented on utils/localization/tests/test_build_locale_nsi.py:
```diff
@@ -0,0 +1,102 @@
+        '!insertmacro MUI_LANGUAGE "German"\n',
+        'LangString test_de1 ${LANG_German} "germantext1"\n',
+        'LangString test_de2 ${LANG_German} "germantext2"\n',
+    ]
```
Nit: Fix missing newline at the end of the file.
```suggestion
    ]

```

patrikjuvonen commented on utils/localization/build_locale_nsi.py:
```diff
@@ -0,0 +1,232 @@
+##############################################################################
+#
+#  PROJECT:     Multi Theft Auto v1.0
```
Nit: Remove version number (we don't use it in source anymore).
```suggestion
#  PROJECT:     Multi Theft Auto
```

patrikjuvonen commented on utils/localization/build_locale_nsi.py:
```diff
@@ -0,0 +1,232 @@
+#  PURPOSE:     Build a multi-language .NSI script based upon input PO files
+#  DEVELOPERS:  Dan Chowdhury <>
+#
+#  Multi Theft Auto is available from http://www.multitheftauto.com/
```
Nit: Use https scheme for links.
```suggestion
#  Multi Theft Auto is available from https://www.multitheftauto.com/
```

patrikjuvonen commented on utils/localization/build_locale_nsi.py:
```diff
@@ -0,0 +1,232 @@
+#
+#  PROJECT:     Multi Theft Auto v1.0
+#  LICENSE:     See LICENSE in the top level directory
+#  FILE:        utils/build_locale_nsi.py
```
Nit: Use correct filepath or alternatively let's remove it. Let's also fix this for https://github.com/multitheftauto/mtasa-blue/blob/5ecc9c731046c4e5392e483d2c6769ab4939a5c9/utils/localization/build_gettext_catalog_nsi.py#L5 and https://github.com/multitheftauto/mtasa-blue/blob/5ecc9c731046c4e5392e483d2c6769ab4939a5c9/utils/localization/build_gettext_catalog.py#L5 please.
```suggestion
#  FILE:        utils/localization/build_locale_nsi.py
```

patrikjuvonen reviewed (changes requested):
Thank you very much @darkdreamingdan for going out your way to upgrade these scripts! This is much appreciated.

I am not fluent in Python so I can't really review the code so much, but it looks clear enough when I eyed it through, so I think it's fine. If there is anyone here fluent in Python, a code review would be appreciated.

I found some nitpick issues regarding some non-code related typos and such, but that's about it.

Code runs well on my Windows machine and it looks to be outputting correct stuff. I haven't extensively tested this (yet).

I wonder if this will run with our [current GitHub Action](https://github.com/multitheftauto/mtasa-blue/blob/5c4d1f4d24061ff06394b98850ad2743f2769255/.github/workflows/rebuild-pots.yaml) out of the box. The goal would be to run these scripts either periodically, or on pull request and push, and create a new pull request for any source changes. This way the source pot remains up-to-date and our [Crowdin project](https://multitheftauto.crowdin.com/multitheftauto) will automatically pull the latest source from Git. This is what our current GitHub Action does at the moment. I believe it is just a matter of upgrading the action's Python version to 3 and updating filepaths if needed. We could also move to Linux instead of Windows on the action for some added efficiency. This is all something I can have a look later in this pull request if that's fine with you, unless you would like to debug and upgrade the action yourself. Either is fine to me.

~~I think the only question from me now is that could we update the xgettext executable to the [latest static version](https://mlocati.github.io/articles/gettext-iconv-windows.html)? I doubt we need the shared version (the one with the DLLs), those could be removed from our repo.~~

All in all great work! :+1:

**Update:** I fixed the nit issues, updated the xgettext executable and rewrote the GitHub actions workflow to take pull requests into account.

patrikjuvonen committed:
Merge branch 'master' into python3_l10n

patrikjuvonen committed:
Add polib to dependencies

Required by build_gettext_catalog_nsi.py

patrikjuvonen committed:
Remove trailing whitespaces and add missing newlines

patrikjuvonen committed:
Fix typo in readme

patrikjuvonen committed:
Fix file path

patrikjuvonen committed:
Remove project version number

patrikjuvonen committed:
Use https scheme

patrikjuvonen committed:
Remove old unused pootle code

patrikjuvonen committed:
Update xgettext.exe to 0.21

patrikjuvonen committed:
Use xgettext as default on non-nt os

patrikjuvonen committed:
Update rebuild-pots workflow to use setup-miniconda action

patrikjuvonen committed:
Upgrade create-pull-request action to v4

patrikjuvonen committed:
Switch to ubuntu-latest

patrikjuvonen committed:
Install gettext

patrikjuvonen committed:
Use the shell from setup-miniconda action example

patrikjuvonen committed:
Run apt-get as sudo

patrikjuvonen committed:
Add pots only if they contain changes beyond POT-Creation-Date

[ci skip]

patrikjuvonen committed:
Only create pull request if there were changes in the previous step

[ci skip]

patrikjuvonen committed:
Turn on echo for debugging step output

patrikjuvonen committed:
Compare against upstream instead

We've already got commits at this point, not staged changes. BRAINFAAART.

[ci skip]

patrikjuvonen committed:
Get rid of all other changes in the end

patrikjuvonen committed:
Try using a placeholder path so create-pull-request wouldn't pick up any extra files

[ci skip]

patrikjuvonen committed:
Delete extra env file as suggested by create-pull-request

[ci skip]

patrikjuvonen committed:
Use boolean style value for clarity

[ci skip]

patrikjuvonen committed:
Use a more comprehensive cleaning method

[ci skip]

patrikjuvonen committed:
Push changes directly instead

Pull requests won't work too well as it requires someone to actively do something about them, causing pull requests and branches to go out of sync.

patrikjuvonen committed:
Name step

[ci skip]

github-actions[bot] committed:
Update client en_US pot

patrikjuvonen committed:
Only run on push

Running on pull_request causes conflicts

patrikjuvonen committed:
Name more steps

[ci skip]

patrikjuvonen committed:
Run pytest

[ci skip]

patrikjuvonen committed:
Add workflow_dispatch trigger
